onnxruntime/core/providers/cpu/nn/tfidfvectorizer.cc

```diff
@@ -490,104 +429,29 @@ Status TfIdfVectorizer::ComputeImpl(OpKernelContext* ctx) const {
     return Status::OK();
   }
 
-  assert((b_dim * C) == total_items);
-
-  const auto max_gram_length = impl.max_gram_length_;
-  const auto max_skip_distance = impl.max_skip_count_ + 1;  // Convert to distance
-  auto start_ngram_size = impl.min_gram_length_;
-  auto const input_data = X->template Data<T>();
-  auto const end_data = input_data + total_items;
-  NgramEntry<T> sample;
-
-  // Treat 1-grams in a special way
-  if (start_ngram_size == 1) {
-    size_t row_num = 0;
-    auto ngram_start = input_data;
-    while (ngram_start < end_data) {
-      auto const ngram_row_end = ngram_start + C;
-      while (ngram_start < ngram_row_end) {
-        sample.Clear();
-        sample.AddItem(*ngram_start);
-        auto hit = impl.PoolFind<T>(sample);
-        if (hit != set_end) {
-          // record frequency
-          auto ngram_id = hit->Id();
-          impl.IncrementCount(ngram_id, row_num, frequencies);
-        }
-        ++ngram_start;
-      }
-      ++row_num;
-      ngram_start = ngram_row_end;
-    }
-    if (++start_ngram_size > max_gram_length) {
-      OutputResult(ctx, B, frequencies);
-      return Status::OK();
+  std::function<void(int32_t)> fn = [this, ctx, C, &frequencies](int32_t row_num) {
+    ComputeImpl(ctx, row_num, C, frequencies);
+  };
+
+  auto tp = ctx->GetOperatorThreadPool();
+  // If tp is nullptr it means we are openmp
+  // and it will run row per thread anyway
+  // if we have our own tp then it does not make sense
+  // to schedule more batches than we have threads + calling thread.
+  if (tp != nullptr) {
+    if (num_rows <= (tp->NumThreads() + 1)) {
```

Comment:
Please see PR #3207

Just wondering, after the change, do you still need the if/else here? Can you just omit the last parameter of BatchParallelFor function if the default value can meet your needs?

Reply:
I will see what I can do.

---
In reply to: [392380226](https://github.com/microsoft/onnxruntime/pull/3197#discussion_r392380226) [](ancestors = 392380226)

Reply:
re-worked for the original code. I let TryBatchParallelFor compute the batch size.

---
In reply to: [392394875](https://github.com/microsoft/onnxruntime/pull/3197#discussion_r392394875) [](ancestors = 392394875,392380226)